### PR TITLE
Specializer in the PIR syntax

### PR DESCRIPTION
**Pull request by GuillaumeGen**
We add a `--transfo` option, which expects a file containing rewriting rules of the form 
```
name : left-hand side ==> right-hand side
```
All those rewriting rules are applied on the system F terms, during the `processDecls` step.
This general mechanism subsumes the `cFoldMapSpecialize` and the `etaIfThenElse`, for which files representing the transformation are provided in this PR.
This is part of the #8 issue. However, it only features PIR to PIR rewriting rules, not PIR to TLA+ rules, which could be the next step.

### CODE REVIEW

File: src/Pirouette/Term/Transformations.hs [hunk 1]
```diff
@@ -351,3 +358,176 @@ chooseHeadCase t ty args fstArg =
...
+    instantiateTy nTy m (R.TyFun a b) =
+      R.TyFun <$> instantiateTy nTy m a <*> instantiateTy nTy m b
+
+    instantiateArg :: (MonadPirouette m)
+                   => Int -> Int -> M.Map String PrtArg -> PrtArg -> m PrtArg
+    instantiateArg nTe nTy m (R.Arg t) = R.Arg <$> instantiate nTe nTy m t
+    instantiateArg nTe nTy m (R.TyArg ty) = R.TyArg <$> instantiateTy nTy m ty
+
+    isHole :: Name -> Maybe String
+    isHole x = head <$> matchRegex (mkRegex "([a-zA-Z][0-9]+)_") (T.unpack $ nameString x)
```
Comment (VictorCMiraldo) [inline src/Pirouette/Term/Transformations.hs:521]: Do we need to bring in an extra dependency for that? Can't we use words starting with an underscore and check whether length x > 1 && `head x == '_'`. Or, if you rather have underscores at the end, `length x > 1 && last x == '_'`

**Review by VictorCMiraldo — COMMENTED** [review]
Great work, thanks @GuillaumeGen !

Besides the small comments I left I was wondering whether we could use a structured format like yaml for the transformations; this way we don't need to write them in a single line, we can have a single file and we can add comments, something like:

```
etaIfThenElse:
  from: [[[{(builtin ifThenElse) A1_} x2_ ] True] False]
  to: x2_

## this transformation is important because XYZ
foldableNilEndo:
  from: ...
  to: ...
```

WDYT?

File: src/Pirouette/Term/Transformations.hs (hunk 1 above)
Comment (GuillaumeGen) [reply src/Pirouette/Term/Transformations.hs:521]: Identifiers in PIR have to start by a letter, so if we keep using the underscore, we have to put it as the last character (or any other position, but the first). I agree that it is a dependency added for a quite restricted use of regular expressions. I can modify this.

**Comment by GuillaumeGen** [comment]
Yes structured files with comments are way better than the prototypic format I have chosen.

**Review by VictorCMiraldo — APPROVED** [review]
Awesome! Thank you @GuillaumeGen

**Pull request merged**